File: Easy/501 - FindModeInBinarySearchTree/Solution.cs
/// Question 501. Find Mode in Binary Search Tree (https://leetcode-cn.com/problems/find-mode-in-binary-search-tree/)

using System.Collections.Generic;
using System;
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */

/// <summary>
/// 执行用时：288 ms, 在所有 C# 提交中击败了 83.67% 的用户
/// 内存消耗：33.7 MB, 在所有 C# 提交中击败了 21.21% 的用户
/// </summary>
public class Solution
{
    private Dictionary<int, int> nodeFrequencies = new Dictionary<int, int>();
    private int highestFrequencies = 0;

    public int[] FindMode(TreeNode root)
    {
        this.TraverseAllNodes(root);
        List<int> modes = new List<int>();

        foreach (KeyValuePair<int, int> keyValuePair in this.nodeFrequencies)
        {
            if (keyValuePair.Value == this.highestFrequencies)
            {
                modes.Add(keyValuePair.Key);
            }
        }

        return modes.ToArray();
    }

    private void TraverseAllNodes(TreeNode root)
    {
        if (root == null)
        {
            return;
        }

        if (nodeFrequencies.ContainsKey(root.val))
        {
            int newFrequency = nodeFrequencies[root.val] + 1;
            nodeFrequencies[root.val] = newFrequency;
            this.highestFrequencies = Math.Max(this.highestFrequencies, newFrequency);
        }
        else
        {
            nodeFrequencies.Add(root.val, 1);
            this.highestFrequencies = Math.Max(this.highestFrequencies, 1);
        }

        TraverseAllNodes(root.left);
        TraverseAllNodes(root.right);
    }
}

File: Easy/501 - FindModeInBinarySearchTree/solution.ts
/**
 * Question 501. Find Mode in Binary Search Tree (https://leetcode-cn.com/problems/find-mode-in-binary-search-tree/)
 *
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * 执行用时：140 ms, 在所有 typescript 提交中击败了 25.00% 的用户
 * 内存消耗：47.8 MB, 在所有 typescript 提交中击败了 100.00% 的用户
 *
 * @param {(TreeNode | null)} root
 * @returns {number[]}
 */
function findMode(root: TreeNode | null): number[] {
  const [frequencyDictionary, highestFrequency] = traverseTree(
    root,
    new Map<number, number>(),
    0
  );

  return Array.from(frequencyDictionary.entries())
    .filter(([_, frequency]) => frequency === highestFrequency)
    .map(([value, _]) => value);
}

function traverseTree(
  root: TreeNode | null,
  frequencyDictionary: Map<number, number>,
  highestFrequency: number
): [Map<number, number>, number] {
  if (!root) {
    return [frequencyDictionary, highestFrequency];
  }

  const rootValue: number = root.val;

  if (frequencyDictionary.has(rootValue)) {
    const newFrequency: number =
      (frequencyDictionary.get(rootValue) as number) + 1;
    highestFrequency = Math.max(newFrequency, highestFrequency);
    frequencyDictionary.set(rootValue, newFrequency);
  } else {
    highestFrequency = Math.max(1, highestFrequency);
    frequencyDictionary.set(rootValue, 1);
  }

  [frequencyDictionary, highestFrequency] = traverseTree(
    root.left,
    frequencyDictionary,
    highestFrequency
  );
  [frequencyDictionary, highestFrequency] = traverseTree(
    root.right,
    frequencyDictionary,
    highestFrequency
  );

  return [frequencyDictionary, highestFrequency];
}
